src/main/java/com/github/packageurl/PackageURL.java

```diff
@@ -247,11 +248,12 @@ private String validateType(final String value) throws MalformedPackageURLExcept
         if (value.charAt(0) >= '0' && value.charAt(0) <= '9') {
             throw new MalformedPackageURLException("The PackageURL type cannot start with a number");
         }
-        final String retVal = value.toLowerCase();
-        if (retVal.chars().anyMatch(c -> !(c == '.' || c == '+' || c == '-'
+        final String retVal = toLowerCase(value);
+        final String invalidChars = retVal.chars().filter(c -> !(c == '.' || c == '+' || c == '-'
                 || (c >= 'a' && c <= 'z')
-                || (c >= '0' && c <= '9')))) {
-            throw new MalformedPackageURLException("The PackageURL type contains invalid characters");
+                || isDigit(c))).mapToObj(c -> String.valueOf((char) c)).collect(Collectors.joining(", "));
```

Comment:
Maybe it would be useful to also refactor `c >= 'a' && c <= 'z'` into a `isLowerAlpha` method. The expression is used 3 times.

Reply:
I did that in a separate pull request #164.